src/relay/backend/graph_runtime_codegen.cc

```diff
@@ -0,0 +1,761 @@
+    }
+    return ret;
+  }
+
+  /*!
+   * \brief Add node to graph
+   *
+   * \param node
+   * \param expr
+   * \return std::vector<_NodeRef>
+   */
+  std::vector<GraphNodeRef> AddNode(GraphNodePtr node, Expr expr) {
+    auto checked_type = expr->checked_type();
+    size_t count = storage_device_map_.count(expr);
+    CHECK_GT(count, 0) << "Expr is not existing in storage plan";
+    auto storage_device_info = storage_device_map_[expr];
+    CHECK_EQ(storage_device_info.size(), 2);
+    // storage
+    std::vector<int64_t> storage_info;
+    for (auto& v : storage_device_info[0]) {
+      storage_info.push_back(v->value);
+    }
+    node->attrs_["storage_id"] = std::move(storage_info);
+    // type
+    std::vector<int64_t> device_types;
+    for (auto& v : storage_device_info[1]) {
+      device_types.push_back(v->value);
+    }
+    size_t num_unknown_devices = std::count(device_types.begin(), device_types.end(), 0);
+    if (num_unknown_devices != 0 && num_unknown_devices != device_types.size()) {
+      LOG(FATAL) << "The graph contains not annotated nodes for "
+                 << "heterogeneous execution. All nodes must be "
+                 << "annotated.";
+    }
+    if (num_unknown_devices == 0) {
+      node->attrs_["device_index"] = device_types;
+    }
+    auto node_id = nodes_.size();
+    nodes_.push_back(node);
+    // Tuple return value, flatten as tuple
+    if (const auto* tuple_type = checked_type.as<TupleTypeNode>()) {
+      std::vector<GraphNodeRef> ret;
+      ShapeVector shape;
+      std::vector<std::string> dtype;
+      for (size_t i = 0; i < tuple_type->fields.size(); ++i) {
+        if (const auto* typ = tuple_type->fields[i].as<TensorTypeNode>()) {
+          ret.push_back(GraphNodeRef(node_id, i));
+          shape.emplace_back(_ShapeToJSON(typ->shape));
+          dtype.emplace_back(DType2String(typ->dtype));
+        } else {
+          LOG(FATAL) << "type " << checked_type->type_key() << " not supported";
+        }
+      }
+      CHECK_EQ(node->Type(), kGraphOpNode);
+      auto op_nd = std::dynamic_pointer_cast<GraphOpNode>(node);
+      op_nd->attrs_["shape"] = shape;
+      op_nd->attrs_["dtype"] = dtype;
+      op_nd->num_outputs_ = tuple_type->fields.size();
+      return ret;
+    }
+    // Normal tensor return type
+    if (const auto* tensor_type = checked_type.as<TensorTypeNode>()) {
+      ShapeVector shape;
+      std::vector<std::string> dtype;
+      shape.emplace_back(_ShapeToJSON(tensor_type->shape));
+      dtype.emplace_back(DType2String(tensor_type->dtype));
+      node->attrs_["shape"] = shape;
+      node->attrs_["dtype"] = dtype;
+    } else {
+      LOG(FATAL) << "type " << checked_type->type_key() << " not supported";
+    }
+    return {GraphNodeRef(node_id, 0)};
+  }
+
+  /*! \brief Visitors */
+  std::unordered_map<Expr, std::vector<GraphNodeRef>, NodeHash, NodeEqual> visitor_cache_;
+
+  std::vector<GraphNodeRef> VisitExpr(const Expr& expr) override {
```

Comment:
Do we need to overload this function?, this should happen by default

Reply:
I think we need.

Reply:
Oh right caching only exists for the Mutator/Visitor